include/arch/arm/aarch64/cpu.h

```diff
@@ -79,4 +119,30 @@
 
 #define GET_EL(_mode)		(((_mode) >> MODE_EL_SHIFT) & MODE_EL_MASK)
 
+#ifdef _ASMLANGUAGE
+/*
+ * macro to support mov of immediate constant to 64 bit register
+ * It will generate instruction sequence of 'mov'/ 'movz' and one
+ * to three 'movk' depending on the immediate value.
+ */
+.macro mov_imm, xreg, imm
+	.if ((\imm) == 0)
+		mov	\xreg, \imm
+	.else
+		.if (((\imm) >> 31) == 0 || ((\imm) >> 31) == 0x1ffffffff)
+			movz    \xreg, (\imm >> 16) & 0xffff, lsl 16
+		.else
+			.if (((\imm) >> 47) == 0 || ((\imm) >> 47) == 0x1ffff)
+				movz    \xreg, (\imm >> 32) & 0xffff, lsl 32
+			.else
+				movz    \xreg, (\imm >> 48) & 0xffff, lsl 48
+				movk    \xreg, (\imm >> 32) & 0xffff, lsl 32
+			.endif
+			movk    \xreg, (\imm >> 16) & 0xffff, lsl 16
+		.endif
+		movk    \xreg, (\imm) & 0xffff, lsl 0
+	.endif
+.endm
+#endif
+
```

Comment:
What is the rationale for moving the `mov_imm` ASM macro from `macro.inc` to `cpu.h`, when all the other ASM macros currently reside in `macro.inc`?

Reply:
Instead of having this one ASM macro inside `cpu.h`, we should probably consider moving the entire `macro.inc` file to `include/arch/arm/aarch64` (or add a separate public ASM macro header).

By the way, the main reason we used the `.inc` extension instead of `.h` was to avoid checkpatch complaints and it also makes sense to use a different extension for a file that contains entirely ASM code.

cc @carlocaione

Reply:
macro.inc does contain all macros  sort of private to arch port.  This instruction alias is something essential for any assembly code to use. So I thought to move it to an existing exported header specific to armv8/arch.

Reply:
`macro.inc` currently has context switch specific macros(`z_arm64_enter_exc`, `z_arm64_exit_exc`) and generic macros(`mov_imm`, `switch_el`).
Current location(`arch/arm/core/aarch64/`) is appropriate for context switch specific macros.
I don't see mov immediate and switch_el macros being used outside of `arch/arm/core/aarch64` unless asm file(*.S) is added at `include/arch/arm/core/aarch64/`.  From that point of view, current `macro.inc` location seems appropriate.

Reply:
> (or add a separate public ASM macro header)

If it is foreseeable that SoC and driver code may find the `mov_imm` macro useful (and possibly others), we should create an additional "public" ASM macro header in `include/arch/arm/aarch64` and add all the public macros there, while keeping the private macros in the arch directory; such that code outside arch can do `#include <arch/arm/aarch64/macro.inc>`.

Also it may be desirable to include the public macro header in the private macro header, so that the arch code does not need to include both.

Reply:
Following what @stephanosio said we should probably leave `z_arm64_{enter,exit}_exc` in `macro.inc` in the `arch/arm/core/aarch64` and move the more generic macros into a more easily accessible `include/arch/arm/aarch64`.

Reply:
Thanks for the suggestions. I will make a patch with a new header, and lets see how that looks.

Reply:
> I don't see mov immediate and switch_el macros being used outside of `arch/arm/core/aarch64`

Currently I will be using `mov_imm` in one place (ref: https://github.com/zephyrproject-rtos/zephyr/pull/24410#discussion_r411279282)  which is arch specific as well as platform/SoC/core specific.  The assembly code will not scale to all sorts of values
if I don't use 'mov_imm'  or if I don't `hand code`  the movz movk sequences.

In future PMU drivers also might need to use a bit of assembly.